addons/gamification/models/ir_http.py

```diff
@@ -0,0 +1,39 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import math
+import time
+from functools import lru_cache
+
+from odoo import models
+from odoo.http import request
+
+EPOCH_DURATION = 300  # 5 min
+
+
+class IrHttp(models.AbstractModel):
+    _inherit = 'ir.http'
+
+    @classmethod
+    def _dispatch(cls, endpoint):
+        result = super()._dispatch(endpoint)
+
+        # This layer avoids fetching the user from db when it is not necessary
+        if request.env.uid:
+            epoch = math.floor(time.time() / EPOCH_DURATION)
+            to_check_update = cls._get_user_last_interacted_window(request.env.uid, epoch)
```

Comment:
wow that's tricky

```py
from functools import lru_cache


@lru_cache
def f():
    return {"a": False}


a = f()
b = f()

print(a, b)  # {'a': False} {'a': False}
a["a"] = True
print(a, b)  # {'a': True} {'a': True}
```

so you code is based on the fact that you modify the cache of `lru_cache` ?

is it really worth it to have something like this just to not fetch the user ?

also because of that, the 5 minutes are duplicated here, and in `action_log_interacted_throttled`

Reply:
I detailed in the commit message the reasons we discussed for having those two gating mechanisms.

Unfortunately, the `user_function` parameter we talked about is in fact the function being decorated when using it as a decorator, not an option to create a custom cache key as available in `cachetools`. 

So we are left with this way of defining the epoch in the caller code, OR we create a "proxy" function for that purpose? Not sure if it is better:

```py
        if request.env.uid:
            to_check_update = cls._get_user_last_interacted_tracker(request)
            ...

    @classmethod
    def _get_user_last_interacted_tracker(cls, request):
        return cls._get_cached_user_last_interacted_tracker(
            request.env.uid, 
            math.floor(time.time() / EPOCH_DURATION)
    )

    @classmethod
    @lru_cache(maxsize=500)
    def _get_cached_user_last_interacted_tracker(cls, uid, epoch):
        return {"used": False}


```
Finally, to be explicitly complete for other readers, the flow in which this setup is used is (also added in a test), 

```py
from functools import lru_cache


@lru_cache
def f(user_id, epoch):
    return {"a": False}


a = f(1, 1)

print(a)  # {'a': False}
a["a"] = True

b = f(1, 1)  # same user and epoch
print(b)  # {'a': True}

c  = f(2, 1)  # other user
print(c)  # {'a': False}

d  = f(1, 2)  # other epoch
print(c)  # {'a': False}
```